### ISSUE TITLE
Add glTF export for maps

### ISSUE THREAD

**Issue by bjorn** [bjorn/tiled#2741]
For general interoperability with 3D engines and tools it would be nice to be able to export a map to a [glTF](https://www.khronos.org/gltf/) file. I think [fx-gltf](https://github.com/jessey-git/fx-gltf) looks like an interesting library to use for this purpose.

**Comment by AniketR10** [comment]
Hey, @bjorn  Aniket here, I am interested in working on this feature, have some experience in C++ (made some command line games and all) and very good proficiency in Javascript/Typescript could you be able to give me some idea what more should i need to know to get started with contribution or is there some issue that i can work on to be able to build the foundation for this.
Thanks!

**Comment by bjorn** [comment]
@AniketR10 You can read a more elaborate description of this feature on the GSoC 2026 wiki page: https://github.com/mapeditor/tiled/wiki/GSoC-2026#gltf-export-for-maps, though I'm sure you're already noticed that. :-)

I have personally not done any further research than the things I wrote there and on this issue, so I'm afraid I can't at the moment provide any additional guidance.

**Comment by kunal649** [comment]
Hii @bjorn,  I've been looking into this. read the glTF 2.0 spec and the existing plugin structure in the repo. Im planning to implement this as a JS plugin (glTF is JSON at its core , and an exporter that runs only once on demand doesn't need C++ complexity). So my starting point will be a minimal skeleton that registers the exporter, iterates tile layers, and outputs a structurally  valid .gltf ->   just enough to open in a validator like [gltf-viewer.donmccurdy.com](https://gltf-viewer.donmccurdy.com/).

I'll work in small incremental PRs -->  tile layer mesh first, then UVs + tileset textures, then flip flags, animations last.
any specific tag/label for PR so u'll easily identify !!?

**Comment by bjorn** [comment]
@kunal649 While I won't stop you from working on this, note that it is a suggested project for this year's GSoC and if you'd like to participate in that we'd rather see you tackle a few other issues first. If you are very effective at coding this feature before GSoC starts, there may not be enough work remaining during the coding period.

Regarding the choice to do this in JS, I pointed to [cgltf](https://github.com/jkuhlmann/cgltf) as possible library to use since it should make it easy to write a standard glTF file (it also supports both binary and JSON formats). I'd also be less worried about performance if this was written as a native plugin. And at least to me, C++ isn't more complex and rather would benefit from a type-checking compiler. But it's not a hard requirement, I'm sure there are benefits to using JS as well.

**Comment by kunal649** [comment]
thats a fair point @bjorn, i can focus on smaller issues first & thanks for calling it out! I would much rather follow your lead on the workflow than learn the hard way by rushing a major feature.
On the language - point taken, I'll look into cgltf and go the C++ route as Type safety does make sense for something where the binary/JSON output has to be precise.

ok so i'll start picking up some smaller issues and will be working on the GSoC proposal for the glTF export in parallel .... so you know where I'm headed.   thanks :-)

**Comment by PrateekSingh070** [comment]
Hi bjorn,

I am working on a GSoC proposal for the glTF export project described in this issue. I have been going through the Tiled codebase (my recent PR #4419 touched compression error reporting across several modules) and I wanted to share my planned approach and ask a couple of questions.

**High-level plan:**
- Each TileLayer becomes a glTF mesh, with one primitive per tileset used in that layer. Tiles are quads (two triangles) with UV coordinates computed from the tileset atlas (accounting for margin, spacing, and tile offset).
- I plan to use cgltf for writing, as you suggested. I have already written a small standalone program that produces a valid .gltf with a textured quad to make sure the library works as expected.
- For tile animations, I am planning to export frame data as `extras` metadata by default, since that is the most portable approach. Morph-target animation would be a stretch goal.
- For map orientations, I will reference the existing renderer math (orthogonalrenderer, isometricrenderer, hexagonalrenderer) to compute vertex positions.

**Questions:**

1. For collection-of-images tilesets, would you prefer the exporter to pack tile images into a runtime-generated atlas at export time, or would it be acceptable to defer that and document it as a limitation initially?

2. For object layers, I am planning to export object data (position, shape, custom properties) as `extras` on glTF nodes. Tile objects would also get rendered as textured quads. Does that match your expectations, or did you have a different approach in mind?

3. Is there a preferred way to expose export options in Tiled's UI for a new plugin? I noticed the TSCN plugin uses `WritableMapFormat` — should I follow the same pattern?

Thanks for any feedback. Happy to adjust the approach based on your input.

— Prateek

**Comment by bjorn** [comment]
@PrateekSingh070 To answer your questions:

* It would be nice to pack image collections into a texture, but it can be a stretch goal.
* Regarding `extras`, I have to admit I don't personally know what this is used for, but it this is used for passing non-rendering information along with the data it sounds appropriate. We'd need to make sure this information can be read out from engines like Godot or Unity (if this isn't practical, a separate JSON file is probably better).
* Indeed currently export formats can't really ship with options. I'd imagine glTF export may need to work more like "Export as Image", which has its own dialog since it also needed a bunch of options.

**Comment by PrateekSingh070** [comment]
Working on a GSoC 2026 proposal for this. Posted the design writeup on the forum so it doesn't bury the issue: https://discourse.mapeditor.org/t/gltf-export-plugin-design-discussion-gsoc-2026-issue-2741/7818

Things I'd like pushback on are the tile-layer → glTF-mesh mapping, how to handle collection-of-images tilesets, and where export options should live in the UI. POC writer is at https://github.com/PrateekSingh070/tiled-gltf-poc (single-quad C demo + 2×2 ortho tile grid in C++ using the real `tileToScreenCoords` math). Scaffold PR #4485 is intentionally a no-op stub while the design's still up for argument.
